res/controllers/Behringer-Extension-scripts.js

```diff
@@ -105,12 +108,13 @@
         components.Button.call(this, options);
     };
     ShiftButton.prototype = deriveFrom(components.Button, {
-        input: function(_channel, _control, value, _status, _group) {
+        inSetValue: function(value) {
             if (value) {
                 this.target.shift();
             } else {
                 this.target.unshift();
             }
+            components.Button.prototype.inSetValue.call(this, value);
```

Comment:
can you explain what the purpose of this change is?

Reply:
I mapped the Shift button to a physical button that has a LED, and noticed that the LED does not work. This was caused by Mixxx not sending anything when the `outKey` control changed, because the input function of `Component` was overridden.

The current change does not override `input` so the original implementation survives, and delegates to the `Button`'s `inSetValue` function so that Shift behaves just as any other button does.

Reply:
Sorry I'm still confused (likely because I have 0 idea how the accompanying library you wrote works). 
The shift button does not seem to be coupled to any CO, so the `outKey` is irrelevant.

you can still overwrite your input implemenation, you just need to send the midi output yourself using `this.send(value)` for example.

Reply:
OK, let me explain it a little bit more detailed; you don't need to understand the whole library for that.

The shift button is first of all a just a normal `Button` that is connected to a user-defined control and additionally calls `shift()` and `unshift()` on some other ("target") component. The user may configure the connected Mixxx control as well as other options (like the button's `type`, for example).

In the [BCR2000 mapping](https://github.com/mixxxdj/mixxx/blob/989f7c8474c2b22dcc31bc00b089f3f09ca3c2b0/res/controllers/Behringer-BCR2000-scripts.js#L25), the connected Mixxx control is configured as `group: [Controls]` & `key: touch_shift`, and the target is the top level `ComponentContainer` that contains all components of the mapping. When the user presses the shift button, `shift()` is called on the top level container and then delegated to all child components. This is effectively a shift for all components of the whole mapping. When the user's skin contains a control for `[Controls],touch_shift`, the shift state is also visualized in the UI.

I didn't use a default for `group` and `key` in the `ShiftButton`'s constructor because it may as well be used for just a group of components or just for a single component. `[Controls],touch_shift` would be misleading in that case.

It's true that I could override `input()` and call `this.send(value)` instead of overriding `inSetValue()`. But when I do so, the `ShiftButton` loses the feature of button types (`push`, `toggle`, `powerWindow`), because this is implemented in `Button.input()`.

Reply:
Ah now I see, thanks for your explanation. I'm sorry I forced you to write this detailed explanation.